File: modules/resize_manager.lua
--- Manages dynamic grid resizing offsets.
-- Allows users to adjust the position of grid lines (separators) for each monitor.
-- @module resize_manager
local hs_timer = hs.timer
local save_debouncer = nil
local resize_manager = {}
local storage = require('modules.storage')

-- State: monitor_id -> { x = { [index] = offset }, y = { [index] = offset } }
-- Offsets are relative to the default grid line position (0.0 - 1.0 scale or pixels? Let's use pixels for simplicity in calculation, or percentage?)
-- Let's use percentage of screen width/height to be resolution independent.
local offsets = {}

-- Configuration
local config = {
    step_size = 0.02 -- 2% adjustment per step
}

--- Loads saved offsets from storage.
function resize_manager.load()
    local data = storage.load('grid_offsets')
    if data then
        offsets = data
    end
end

--- Saves offsets to storage.
function resize_manager.save()
    storage.save('grid_offsets', offsets)
end

--- Gets the offset for a specific grid line.
---@param monitor_id string
---@param axis string "x" (cols) or "y" (rows)
---@param index number The index of the grid line (1 to N-1)
---@return number offset The offset in percentage (e.g., 0.05 for +5%)
function resize_manager.get_offset(monitor_id, axis, index)
    if offsets[monitor_id] and offsets[monitor_id][axis] then
        return offsets[monitor_id][axis][index] or 0
    end
    return 0
end

--- Adjusts a grid line.
---@param monitor_id string
---@param axis string "x" or "y"
---@param index number The index of the grid line.
---@param delta number The amount to change (e.g., 1 for +step, -1 for -step)
function resize_manager.adjust(monitor_id, axis, index, delta)
    if not offsets[monitor_id] then
        offsets[monitor_id] = {
            x = {},
            y = {}
        }
    end
    if not offsets[monitor_id][axis] then
        offsets[monitor_id][axis] = {}
    end

    local current = offsets[monitor_id][axis][index] or 0
    local new_val = current + (delta * config.step_size)

    -- Clamp
    if new_val > 0.4 then
        new_val = 0.4
    end
    if new_val < -0.4 then
        new_val = -0.4
    end

    offsets[monitor_id][axis][index] = new_val

    -- ONLY update the UI instantly. Defer the disk write.
    if save_debouncer then
        save_debouncer:stop()
    end
    save_debouncer = hs_timer.delayed.new(1.0, function()
        resize_manager.save()
        save_debouncer = nil
    end)
    save_debouncer:start()
end

--- Resets offsets for a monitor.
---@param monitor_id string
function resize_manager.reset(monitor_id)
    if offsets[monitor_id] then
        offsets[monitor_id] = nil
        resize_manager.save()
    end
end

--- Initializes the manager.
function resize_manager.init()
    resize_manager.load()
end

return resize_manager

File: Sources/ZTCore/ResizeManager.swift
// ResizeManager.swift — port of modules/resize_manager.lua. Holds per-monitor grid-line
// offsets (percentage of screen size, one per axis/index), adjustable in ±2% steps and
// clamped to ±40%. Supplies ZoneCalculator's OffsetProvider and persists via Storage.
//
// The Lua module's debounced save (hs.timer.delayed) is system glue; here saving is
// explicit (the system layer debounces). Pure logic + serialization only.

public final class ResizeManager {

    public static let stepSize = 0.02
    public static let clampLimit = 0.4
    public static let storageKey = "grid_offsets"

    /// On-disk shape: per monitor, an offsets-by-index map per axis (index keys are strings
    /// for JSON). Matches what grid_offsets.json holds.
    public struct MonitorOffsets: Codable, Equatable {
        public var x: [String: Double]
        public var y: [String: Double]
        public init(x: [String: Double] = [:], y: [String: Double] = [:]) { self.x = x; self.y = y }
    }

    // monitor -> axis ("x"/"y") -> grid-line index -> offset
    private var offsets: [String: [String: [Int: Double]]] = [:]

    public init() {}

    public func getOffset(monitor: String, axis: String, index: Int) -> Double {
        offsets[monitor]?[axis]?[index] ?? 0
    }

    public func adjust(monitor: String, axis: String, index: Int, delta: Double) {
        let current = offsets[monitor]?[axis]?[index] ?? 0
        var newVal = current + delta * Self.stepSize
        newVal = min(Self.clampLimit, max(-Self.clampLimit, newVal))
        offsets[monitor, default: [:]][axis, default: [:]][index] = newVal
    }

    public func reset(monitor: String) {
        offsets[monitor] = nil
    }

    /// A ZoneCalculator OffsetProvider bound to a specific monitor.
    public func offsetProvider(monitor: String) -> ZoneCalculator.OffsetProvider {
        { [weak self] axis, index in self?.getOffset(monitor: monitor, axis: axis, index: index) ?? 0 }
    }

    // MARK: - Serialization

    public func snapshot() -> [String: MonitorOffsets] {
        var out: [String: MonitorOffsets] = [:]
        for (monitor, axes) in offsets {
            var mo = MonitorOffsets()
            for (index, value) in axes["x"] ?? [:] { mo.x[String(index)] = value }
            for (index, value) in axes["y"] ?? [:] { mo.y[String(index)] = value }
            out[monitor] = mo
        }
        return out
    }

    public func restore(_ data: [String: MonitorOffsets]) {
        var loaded: [String: [String: [Int: Double]]] = [:]
        for (monitor, mo) in data {
            var x: [Int: Double] = [:], y: [Int: Double] = [:]
            for (k, v) in mo.x { if let i = Int(k) { x[i] = v } }
            for (k, v) in mo.y { if let i = Int(k) { y[i] = v } }
            loaded[monitor] = ["x": x, "y": y]
        }
        offsets = loaded
    }

    public func load(from storage: Storage) {
        if let data = storage.load(Self.storageKey, as: [String: MonitorOffsets].self) {
            restore(data)
        }
    }

    @discardableResult
    public func save(to storage: Storage) -> Bool {
        storage.save(Self.storageKey, snapshot())
    }
}
